internal/completion/complete.go

```diff
@@ -255,23 +209,161 @@ func NewCompleteCmd(settings *cli.EnvSettings, out io.Writer) *cobra.Command {
+			if subCmd.IsAvailableCommand() && strings.HasPrefix(subCmd.Name(), toComplete) {
+				comp := subCmd.Name()
+				if includeDesc {
+					comp = fmt.Sprintf("%s\t%s", comp, subCmd.Short)
+				}
+				completions = append(completions, comp)
+			}
+		}
+
+		// Always complete ValidArgs, even if we are completing a subcommand name.
+		// This is for commands that have both subcommands and validArgs.
+		for _, validArg := range finalCmd.ValidArgs {
+			if strings.HasPrefix(validArg, toComplete) {
+				completions = append(completions, validArg)
+			}
+		}
+
+		// Always let the logic continue to add any ValidArgsFunction completions,
+		// even if we already found other completions already.
+		// This is for commands that have subcommands and/or validArgs but also
+		// specify a ValidArgsFunction.
+	}
+
+	// Parse the flags and extract the arguments to prepare for calling the completion function
+	if err = finalCmd.ParseFlags(finalArgs); err != nil {
+		return completions, BashCompDirectiveDefault, fmt.Errorf("Error while parsing flags from args %v: %s", finalArgs, err.Error())
+	}
+
+	// We only remove the flags from the arguments if DisableFlagParsing is not set.
+	// This is important for helm plugins, which need to receive all flags.
+	// The plugin completion code will do its own flag parsing.
+	if !finalCmd.DisableFlagParsing {
+		finalArgs = finalCmd.Flags().Args()
+	}
+
+	// Find completion function for the flag or command
+	var key interface{}
+	var nameStr string
+	if flag != nil {
+		key = flag
+		nameStr = flag.Name
+	} else {
+		key = finalCmd
+		nameStr = finalCmd.CommandPath()
+	}
+	completionFn, ok := validArgsFunctions[key]
+	if !ok {
+		return completions, BashCompDirectiveDefault, fmt.Errorf("Go custom completion not supported/needed for flag or command: %s", nameStr)
+	}
+
+	comps, directive := completionFn(finalCmd, finalArgs, toComplete)
+	completions = append(completions, comps...)
+	return completions, directive, nil
+}
+
+func getFlagNameCompletions(flag *pflag.Flag, toComplete string, includeDesc bool) []string {
+	if nonCompletableFlag(flag) {
+		return []string{}
+	}
+
+	var completions []string
+	comp := "--" + flag.Name
+	if strings.HasPrefix(comp, toComplete) {
+		// Flag without the =
+		completions = append(completions, comp)
+
+		if len(flag.NoOptDefVal) == 0 {
+			// Flag requires a value, so it can be suffixed with =
+			comp += "="
+			completions = append(completions, comp)
+		}
+	}
+
+	comp = "-" + flag.Shorthand
+	if len(flag.Shorthand) > 0 && strings.HasPrefix(comp, toComplete) {
+		completions = append(completions, comp)
+	}
+
+	// Add documentation if requested
+	if includeDesc {
+		for idx, comp := range completions {
+			completions[idx] = fmt.Sprintf("%s\t%s", comp, flag.Usage)
+		}
+	}
+	return completions
+}
+
 func isFlag(arg string) bool {
 	return len(arg) > 0 && arg[0] == '-'
 }
 
+func nonCompletableFlag(flag *pflag.Flag) bool {
+	return flag.Hidden || len(flag.Deprecated) > 0
```

Comment:
Instead of a length check, can we check `flag.Deprecated` against the empty string here? It's a little clearer to the reader what we're trying to check for

Reply:
Done